c/strOrFile.c

```diff
@@ -0,0 +1,168 @@
+		else if ((inputStr != NULL) && (strlen(inputStr) > 0))
+		{
+			theStrOrFile->inputStr = inputStr;
+			theStrOrFile->inputStrPos = 0;
+		}
+	}
+	
+	return theStrOrFile;
+}
+
+/********************************************************************
+ sf_getc()
+ If strOrFileP has FILE pointer to input file, calls getc().
+ If strOrFileP has input string, returns character at inputStrPos and
+ increments inputStrPos.
+ ********************************************************************/
+char sf_getc(strOrFileP theStrOrFile)
+{
+	char theChar = '\0';
+
+	if (theStrOrFile != NULL)
+	{
+		if (theStrOrFile->inputFile != NULL)
+			theChar = getc(theStrOrFile->inputFile);
+		else if (theStrOrFile->inputStr != NULL)
+		{
+			if ((theStrOrFile->inputStr + theStrOrFile->inputStrPos)[0] == '\0')
+				return EOF;
+
+			theChar = theStrOrFile->inputStr[theStrOrFile->inputStrPos++];
+		} 
+	}
+	
+	return theChar;
+}
+
+/********************************************************************
+ sf_ungetc()
+ Order of parameters matches stdio ungetc().
+
+ If strOrFileP has FILE pointer to input file, calls ungetc().
+ If strOrFileP has input string, decrements inputStrPos and returns
+ character at inputStrPos.
+
+ Like ungetc() in stdio, on success theChar is returned. On failure, EOF
+ is returned. 
+ ********************************************************************/
+char sf_ungetc(char theChar, strOrFileP theStrOrFile)
+{
+	char charToReturn = EOF;
+
+	if (theStrOrFile != NULL)
+	{
+		if (theStrOrFile->inputFile != NULL)
+			charToReturn = ungetc(theChar, theStrOrFile->inputFile);
+		// Don't want to ungetc to an index before theStrOrFile->inputStr start
+		else if (theStrOrFile->inputStr != NULL)
+		{
+			if (theStrOrFile->inputStrPos <= 0)
+				return EOF;
+
+			// Decrement inputStrPos, then replace the character in inputStr at that position with theChar
+			charToReturn = theStrOrFile->inputStr[--(theStrOrFile->inputStrPos)] = theChar;
+		}
+	}
+
+	return charToReturn;
+}
+
+/********************************************************************
+ sf_fgets()
+ Order of parameters matches stdio fgets().
+
+ First param is the string to populate, second param
+ is the max number of characters to read, and third param is the pointer to the
+ string-or-file container from which we wish to read count characters.
+
+ Like fgets() in stdio, this function doesn't check that enough memory
+ is allocated for str to contain (count - 1) characters.
+
+ Like fgets() in stdio, on success the pointer to the buffer is returned.
+ On failure, NULL is returned.
+ ********************************************************************/
+char * sf_fgets(char *str, int count, strOrFileP theStrOrFile)
+{
+int numCharsRead = -1;
+
+	if (str == NULL || count < 0 || theStrOrFile == NULL)
+		return NULL;
+	
+	if (theStrOrFile->inputFile != NULL)
+	{
+		fgets(str, count, theStrOrFile->inputFile);
+		return str;
+	}
+	else if (theStrOrFile->inputStr != NULL)
```

Comment:
I think you need to add to this condition `&& theStrOrFile->inputStr[0] != '\0'` because we want `sf_fgets()` to return NULL when we reach the string version of EOF, which is the null terminator.